File: services/agent-api/representative_timetable.py
"""Deterministic search over the two private representative timetables."""
from __future__ import annotations

import gzip
import json
import unicodedata
from typing import Any


class TimetableSearchError(ValueError):
    pass


_cache: dict[str, tuple[str, dict[str, Any]]] = {}


def search(
    s3_client: Any,
    bucket: str,
    prefix: str,
    request: dict[str, Any],
) -> dict[str, Any]:
    kind = request.get("timetableKind")
    query = request.get("query")
    mode = request.get("mode")
    target_time = request.get("targetTimeMinutes")
    limit = request.get("limit", 5)
    if kind not in {"weekday", "weekend_holiday"}:
        raise TimetableSearchError("timetableKindが不正です。")
    if not isinstance(query, str) or not query.strip() or len(query) > 200:
        raise TimetableSearchError("queryが不正です。")
    if mode not in {"active", "arrivals", "departures"}:
        raise TimetableSearchError("modeが不正です。")
    if target_time is not None and (
        not isinstance(target_time, (int, float))
        or isinstance(target_time, bool)
        or not 0 <= target_time <= 2_160
    ):
        raise TimetableSearchError("targetTimeMinutesが不正です。")
    if not isinstance(limit, int) or isinstance(limit, bool):
        raise TimetableSearchError("limitが不正です。")
    limit = max(1, min(5, limit))

    timetable = _load(s3_client, bucket, prefix, kind)
    normalized_query = _normalize(query)
    trains = timetable.get("trains")
    if not isinstance(trains, list):
        raise RuntimeError("Representative timetable has no trains array")
    station_names = {
        _normalize(stop.get("station_name"))
        for train in trains if isinstance(train, dict)
        for stop in train.get("stops", []) if isinstance(stop, dict)
        if isinstance(stop.get("station_name"), str)
        and _normalize(stop["station_name"]) in normalized_query
    }
    service_types = {
        _normalize(train.get("service_type"))
        for train in trains if isinstance(train, dict)
        if isinstance(train.get("service_type"), str)
        and _normalize(train["service_type"]) in normalized_query
    }
    train_names = {
        _normalize(train.get("train_name"))
        for train in trains if isinstance(train, dict)
        if isinstance(train.get("train_name"), str) and train.get("train_name")
        and _normalize(train["train_name"]) in normalized_query
    }
    train_numbers = {
        _normalize(train.get("train_no"))
        for train in trains if isinstance(train, dict)
        if isinstance(train.get("train_no"), str)
        and len(_normalize(train["train_no"])) >= 2
        and _normalize(train["train_no"]) in normalized_query
    }
    if not any((station_names, service_types, train_names, train_numbers)):
        return _response(timetable, kind, mode, target_time, [], 0)

    ranked: list[tuple[float, dict[str, Any]]] = []
    for train in trains:
        if not isinstance(train, dict) or not _train_matches(
            train, station_names, service_types, train_names, train_numbers
        ):
            continue
        matching_stops = _matching_stops(train, station_names, mode, target_time)
        if mode == "active" and target_time is not None:
            times = [
                stop.get("route_time_minutes")
                for stop in train.get("stops", []) if isinstance(stop, dict)
                if isinstance(stop.get("route_time_minutes"), (int, float))
            ]
            if not times or not min(times) <= target_time <= max(times):
                continue
        elif mode != "active" and not matching_stops:
            continue
        nearest = (
            min(
                (abs(stop["routeTimeMinutes"] - target_time) for stop in matching_stops),
                default=0,
            )
            if target_time is not None else 0
        )
        ranked.append((nearest, {
            "trainNumber": train.get("train_no", ""),
            "serviceType": train.get("service_type", ""),
            "trainName": train.get("train_name", ""),
            "origin": train.get("origin_station", ""),
            "destination": train.get("destination_station", ""),
            "matchingStops": matching_stops[:4],
        }))
    ranked.sort(key=lambda item: (item[0], item[1]["trainNumber"]))
    return _response(
        timetable, kind, mode, target_time,
        [item[1] for item in ranked[:limit]], len(ranked),
    )


def _load(s3_client: Any, bucket: str, prefix: str, kind: str) -> dict[str, Any]:
    key = f"{prefix.strip('/')}/{kind}.json.gz"
    etag = str(s3_client.head_object(Bucket=bucket, Key=key).get("ETag", ""))
    cached = _cache.get(kind)
    if cached is not None and cached[0] == etag:
        return cached[1]
    body = s3_client.get_object(Bucket=bucket, Key=key)["Body"].read()
    value = json.loads(gzip.decompress(body))
    if not isinstance(value, dict) or value.get("schema_version") != "ai-timetable-v1":
        raise RuntimeError("Representative timetable schema is invalid")
    _cache[kind] = (etag, value)
    return value


def _train_matches(
    train: dict[str, Any], stations: set[str], service_types: set[str],
    train_names: set[str], train_numbers: set[str],
) -> bool:
    train_stations = {
        _normalize(stop.get("station_name"))
        for stop in train.get("stops", []) if isinstance(stop, dict)
        if isinstance(stop.get("station_name"), str)
    }
    return (
        (not stations or bool(stations & train_stations))
        and (not service_types or _normalize(train.get("service_type")) in service_types)
        and (not train_names or _normalize(train.get("train_name")) in train_names)
        and (not train_numbers or _normalize(train.get("train_no")) in train_numbers)
    )


def _matching_stops(
    train: dict[str, Any], stations: set[str], mode: str, target_time: float | None
) -> list[dict[str, Any]]:
    result = []
    event_text = "着" if mode == "arrivals" else "発" if mode == "departures" else None
    for stop in train.get("stops", []):
        if not isinstance(stop, dict):
            continue
        station = stop.get("station_name")
        route_time = stop.get("route_time_minutes")
        if not isinstance(station, str) or not isinstance(route_time, (int, float)):
            continue
        if stations and _normalize(station) not in stations:
            continue
        if event_text is not None and event_text not in str(stop.get("event", "")):
            continue
        if target_time is not None and abs(route_time - target_time) > 30:
            continue
        result.append({
            "stationName": station,
            "event": stop.get("event", ""),
            "routeTimeMinutes": route_time,
        })
    return result


def _response(
    timetable: dict[str, Any], kind: str, mode: str,
    target_time: float | None, matches: list[dict[str, Any]], total: int,
) -> dict[str, Any]:
    return {
        "timetableKind": kind,
        "serviceDate": timetable.get("service_date"),
        "mode": mode,
        "targetTimeMinutes": target_time,
        "totalMatchCount": total,
        "matches": matches,
    }


def _normalize(value: Any) -> str:
    if not isinstance(value, str):
        return ""
    return unicodedata.normalize("NFKC", value).strip().replace("ヶ", "ケ")

File: backend/agent-api/src/usecases/representative-timetable.ts
import { type JsonObject, RequestError } from "../contracts/agent-request.js";
import type { AgentOperation } from "../ports/agent-operation.js";
import type { RepresentativeTimetableKind, RepresentativeTimetableRepository } from "../ports/operation-data.js";

type SearchMode = "active" | "arrivals" | "departures";

export function createRepresentativeTimetableOperation(repository: RepresentativeTimetableRepository): AgentOperation {
  return async (request) => ({ body: await searchRepresentativeTimetable(repository, request) });
}

export async function searchRepresentativeTimetable(
  repository: RepresentativeTimetableRepository,
  request: JsonObject,
): Promise<JsonObject> {
  const kind = request.timetableKind;
  const query = request.query;
  const mode = request.mode;
  const targetTime = request.targetTimeMinutes;
  const rawLimit = request.limit ?? 5;
  if (kind !== "weekday" && kind !== "weekend_holiday") throw new RequestError(400, "timetableKindが不正です。");
  if (typeof query !== "string" || !query.trim() || query.length > 200) throw new RequestError(400, "queryが不正です。");
  if (mode !== "active" && mode !== "arrivals" && mode !== "departures") throw new RequestError(400, "modeが不正です。");
  if (targetTime !== undefined && targetTime !== null && (typeof targetTime !== "number" || !Number.isFinite(targetTime) || targetTime < 0 || targetTime > 2_160)) throw new RequestError(400, "targetTimeMinutesが不正です。");
  if (!Number.isInteger(rawLimit)) throw new RequestError(400, "limitが不正です。");
  const limit = Math.max(1, Math.min(5, rawLimit as number));
  const timetable = await repository.load(kind);
  if (!Array.isArray(timetable.trains)) throw new Error("Representative timetable has no trains array");
  const trains = timetable.trains.filter(isRecord);
  const normalizedQuery = normalize(query);
  const stations = matchingValues(trains.flatMap((train) => Array.isArray(train.stops) ? train.stops.filter(isRecord).map((stop) => stop.station_name) : []), normalizedQuery);
  const serviceTypes = matchingValues(trains.map((train) => train.service_type), normalizedQuery);
  const trainNames = matchingValues(trains.map((train) => train.train_name).filter((value) => normalize(value) !== ""), normalizedQuery);
  const trainNumbers = matchingValues(trains.map((train) => train.train_no).filter((value) => normalize(value).length >= 2), normalizedQuery);
  if (stations.size + serviceTypes.size + trainNames.size + trainNumbers.size === 0) return response(timetable, kind, mode, targetTime, [], 0);
  const ranked = trains.flatMap((train) => {
    if (!trainMatches(train, stations, serviceTypes, trainNames, trainNumbers)) return [];
    const stops = matchingStops(train, stations, mode, typeof targetTime === "number" ? targetTime : null);
    if (mode === "active" && typeof targetTime === "number") {
      const times = (Array.isArray(train.stops) ? train.stops : []).filter(isRecord).map((stop) => stop.route_time_minutes).filter((value): value is number => typeof value === "number");
      if (times.length === 0 || targetTime < Math.min(...times) || targetTime > Math.max(...times)) return [];
    } else if (mode !== "active" && stops.length === 0) return [];
    const nearest = typeof targetTime === "number" && stops.length > 0
      ? Math.min(...stops.map((stop) => Math.abs((stop.routeTimeMinutes as number) - targetTime)))
      : 0;
    return [{ nearest, value: { trainNumber: stringValue(train.train_no), serviceType: stringValue(train.service_type), trainName: stringValue(train.train_name), origin: stringValue(train.origin_station), destination: stringValue(train.destination_station), matchingStops: stops.slice(0, 4) } }];
  }).sort((a, b) => a.nearest - b.nearest || String(a.value.trainNumber).localeCompare(String(b.value.trainNumber)));
  return response(timetable, kind, mode, targetTime, ranked.slice(0, limit).map(({ value }) => value), ranked.length);
}

function matchingValues(values: unknown[], query: string): Set<string> {
  return new Set(values.filter((value): value is string =>
    typeof value === "string" && normalize(value).length > 0 && query.includes(normalize(value))
  ).map(normalize));
}

function trainMatches(train: JsonObject, stations: Set<string>, serviceTypes: Set<string>, trainNames: Set<string>, trainNumbers: Set<string>): boolean {
  const trainStations = new Set((Array.isArray(train.stops) ? train.stops : []).filter(isRecord).map((stop) => normalize(stop.station_name)));
  return (stations.size === 0 || [...stations].some((station) => trainStations.has(station))) &&
    (serviceTypes.size === 0 || serviceTypes.has(normalize(train.service_type))) &&
    (trainNames.size === 0 || trainNames.has(normalize(train.train_name))) &&
    (trainNumbers.size === 0 || trainNumbers.has(normalize(train.train_no)));
}

function matchingStops(train: JsonObject, stations: Set<string>, mode: SearchMode, targetTime: number | null): JsonObject[] {
  const eventText = mode === "arrivals" ? "着" : mode === "departures" ? "発" : undefined;
  return (Array.isArray(train.stops) ? train.stops : []).filter(isRecord).flatMap((stop) => {
    if (typeof stop.station_name !== "string" || typeof stop.route_time_minutes !== "number") return [];
    if (stations.size > 0 && !stations.has(normalize(stop.station_name))) return [];
    if (eventText && !String(stop.event ?? "").includes(eventText)) return [];
    if (targetTime !== null && Math.abs(stop.route_time_minutes - targetTime) > 30) return [];
    return [{ stationName: stop.station_name, event: stringValue(stop.event), routeTimeMinutes: stop.route_time_minutes }];
  });
}

function response(timetable: JsonObject, kind: RepresentativeTimetableKind, mode: SearchMode, targetTime: unknown, matches: JsonObject[], total: number): JsonObject {
  return { timetableKind: kind, serviceDate: timetable.service_date, mode, targetTimeMinutes: targetTime ?? null, totalMatchCount: total, matches };
}

function normalize(value: unknown): string { return typeof value === "string" ? value.normalize("NFKC").trim().replaceAll("ヶ", "ケ") : ""; }
function stringValue(value: unknown): string { return typeof value === "string" ? value : ""; }
function isRecord(value: unknown): value is JsonObject { return typeof value === "object" && value !== null && !Array.isArray(value); }
